lib/_tls_wrap.js

```diff
@@ -37,211 +37,243 @@ const tls_wrap = process.binding('tls_wrap');
-    self._emitTLSError(err);
+    owner._emitTLSError(err);
   }
 }
 
+function emitSessionAttackError(self) {
+  self._emitTLSError(new errors.Error('ERR_TLS_SESSION_ATTACK'));
+}
 
 function onhandshakedone() {
-  // for future use
   debug('onhandshakedone');
-  this._finishInit();
+
+  const owner = this.owner;
+
+  // `newSession` callback wasn't called yet
+  if (owner._newSessionPending) {
+    owner._securePending = true;
+    return;
+  }
+
+  owner._finishInit();
 }
 
 
-function loadSession(self, hello, cb) {
+function loadSession(hello) {
+  const owner = this.owner;
+
   var once = false;
   function onSession(err, session) {
     if (once)
-      return cb(new errors.Error('ERR_MULTIPLE_CALLBACK'));
+      return loadSessionDone(new errors.Error('ERR_MULTIPLE_CALLBACK'), owner);
     once = true;
 
     if (err)
-      return cb(err);
+      return loadSessionDone(err, owner);
 
-    if (!self._handle)
-      return cb(new errors.Error('ERR_SOCKET_CLOSED'));
+    if (owner._handle === null)
+      return loadSessionDone(new errors.Error('ERR_SOCKET_CLOSED'), owner);
 
-    self._handle.loadSession(session);
-    cb(null);
+    owner._handle.loadSession(session);
+    loadSessionDone(null, owner);
   }
 
   if (hello.sessionId.length <= 0 ||
       hello.tlsTicket ||
-      self.server &&
-      !self.server.emit('resumeSession', hello.sessionId, onSession)) {
-    cb(null);
+      owner.server &&
+      !owner.server.emit('resumeSession', hello.sessionId, onSession)) {
+    loadSessionDone(null, owner);
   }
 }
 
+function loadSessionDone(err, self) {
+  if (err !== null)
+    return self.destroy(err);
 
-function loadSNI(self, servername, cb) {
-  if (!servername || !self._SNICallback)
-    return cb(null);
+  self._handle.endParser();
+}
 
-  var once = false;
-  self._SNICallback(servername, function(err, context) {
+
+function loadSNI(info) {
+  const owner = this.owner;
+  const servername = info.servername;
+  if (!servername || !owner._SNICallback)
+    return requestOCSP(null, owner, info);
+
+  let once = false;
+  owner._SNICallback(servername, (err, context) => {
     if (once)
-      return cb(new errors.Error('ERR_MULTIPLE_CALLBACK'));
+      return requestOCSP(new errors.Error('ERR_MULTIPLE_CALLBACK'), owner);
     once = true;
 
     if (err)
-      return cb(err);
+      return requestOCSP(err, owner);
 
-    if (!self._handle)
-      return cb(new errors.Error('ERR_SOCKET_CLOSED'));
+    if (owner._handle === null)
+      return requestOCSP(new errors.Error('ERR_SOCKET_CLOSED'), owner);
 
     // TODO(indutny): eventually disallow raw `SecureContext`
     if (context)
-      self._handle.sni_context = context.context || context;
+      owner._handle.sni_context = context.context || context;
 
-    cb(null, self._handle.sni_context);
+    requestOCSP(null, owner, info);
   });
 }
 
 
-function requestOCSP(self, hello, ctx, cb) {
+function requestOCSP(err, self, hello) {
```

Comment:
The calling functions calls `hello` info. This is somewhat inconsistent. 

I see why you changed `owner` to `self` but I am also not very happy about that naming either. Right now I am not certain what would be best.